### PR TITLE
Add OAuth config support

### PR DESCRIPTION
**Pull request by rinzool**
## Why
The purpose of this PR is to add the possibility to use [OAuth M2M](https://docs.databricks.com/en/dev-tools/authentication-oauth.html) login with databricks plugin
See: https://github.com/turbot/steampipe-plugin-databricks/issues/7 

## How
* Update README, `docs/index.html`,  `config/databricks.spc` file 
* Update `databricksConfig` object to add `ClientId` and `ClientSecret`
* Update `service.go` to set `DATABRICKS_CLIENT_ID` and `DATABRICKS_CLIENT_SECRET` environment variables

## Tests
* [x] Use a client_id/client_secret and see that it works

### CODE REVIEW

**Comment by rinzool** [comment]
> Hey @rinzool , thanks for opening this PR!
> 
> We're actually considering switching how we set credentials in the client (#8), in particular moving away from setting env vars and passing in the creds specified either in the config args and/or env vars.
> 
> This is work we're planning to pick up next week, so we may hold off on reviewing this PR until we get the changes in for the existing code, and then we can share the updated code so this PR can follow the same conventions.
> 
> If you have any questions on that approach, please let us know!

Ok nice idea :+1: 
Keep me touch when it's done so I can update my PR :wink: (And if you want any external review too :smiley:)

**Comment by rinzool** [comment]
Hi @cbruno10, any update about when https://github.com/turbot/steampipe-plugin-databricks/issues/8 will be closed? I see that a PR is opened

**Comment by cbruno10** [comment]
Hey @rinzool , that should be closed by end of this week, apologies for the delay!

**Comment by misraved** [comment]
Apologies for the delay @rinzool, we will be closing the https://github.com/turbot/steampipe-plugin-databricks/pull/10 PR early next week (Thanks a ton for reviewing the PR 🎉 ).

Post that we can get your changes in 👍.

**Comment by rinzool** [comment]
> Apologies for the delay @rinzool, we will be closing the #10 PR early next week (Thanks a ton for reviewing the PR 🎉 ).
> 
> Post that we can get your changes in 👍.

Ok nice!
I will rebase on #10 branch so my PR will be ready directly after :ok_hand:

**Comment by rinzool** [comment]
Hi @misraved, any updates about #10? :pray:

**Comment by rinzool** [comment]
Hi @misraved, any updates?
I really need this feature to use OAuth to interact with databricks plugin :pray:

**Comment by misraved** [comment]
Hi @rinzool, sorry for the delay!!

Could you please resolve the merge conflicts, so that we can get the final review done?

**Comment by rinzool** [comment]
Hi @misraved 
I see that #10 has an approval, do you think it will soon be merged so I can rebase this PR and ask for reviews? :pray:

**Comment by bigdatasourav** [comment]
Hey @rinzool, The PR https://github.com/turbot/steampipe-plugin-databricks/pull/10 is merged. Could you please resolve the conflicts and prepare this PR for review?

**Comment by rinzool** [comment]
> Hey @rinzool, The PR #10 is merged. Could you please resolve the conflicts and prepare this PR for review?

Yep it's done. I tested it and it works perfectly :ok_hand:

**Comment by rinzool** [comment]
Hi @bigdatasourav, what do you think about this PR now the conflict has been resolved? :pray:

File: databricks/service.go [hunk 1]
```diff
@@ -56,6 +56,11 @@ func getAccountClientUncached(ctx context.Context, d *plugin.QueryData, h *plugi
 			config.Token = *databricksConfig.AccountToken
 		}
 
+                if databricksConfig.ClientID != nil && databricksConfig.ClientSecret != nil {
```
Comment (bigdatasourav) [inline databricks/service.go:59]: We need to verify and establish the values of ClientID and ClientSecret if they are being set through the environment variables DATABRICKS_CLIENT_ID and DATABRICKS_CLIENT_SECRET.

Comment (rinzool) [reply databricks/service.go:59]: If environment variables exists, the databricks sdk should automatically use them if I understand the documentation: https://pkg.go.dev/github.com/databricks/databricks-sdk-go#readme-default-authentication-flow
Currently in the `service.go` file we don't use environment variables for token/username/password. I think it should be the same for client id/client secret, no?

File: docs/index.md [hunk 2]
```diff
@@ -290,6 +298,23 @@ connection "databricks_user1-workspace" {
...
+
+Configuration to query Databricks workspace by using [OAuth for service principals](https://docs.databricks.com/en/dev-tools/authentication-oauth.html).
+
+#### databricks.spc
+
+```hcl
+connection "databricks_user1-account" {
+  plugin  = "databricks"
+  account_id = "abcdd0f81-9be0-4425-9e29-3a7d96782373"
+  workspace_host = "https://dbc-a1b2c3d4-e6f7.cloud.databricks.com/"
```
Comment (bigdatasourav) [inline docs/index.md:311]: Is this also applicable to the AccountHost?

Comment (rinzool) [reply docs/index.md:311]: Yes it is.
According to [Databricks documentation](https://docs.databricks.com/en/dev-tools/auth/oauth-u2m.html) one need to set
`workspace_host = "https://accounts.cloud.databricks.com"` and to define `account_id` in that case

File: databricks/service.go [hunk 3]
```diff
@@ -122,6 +127,11 @@ func getWorkspacetClientUncached(ctx context.Context, d *plugin.QueryData, h *pl
 			config.Token = *databricksConfig.WorkspaceToken
 		}
 
+                if databricksConfig.ClientID != nil && databricksConfig.ClientSecret != nil {
```
Comment (bigdatasourav) [inline databricks/service.go:130]: What is the order we are maintaining here? Below is the order we are following. Now should we check for ClientID and ClientSecret before Username and Password?
 
// Account ID is required for a common column
// Check for profile and config file first
// If not using a profile and config file, check for token
// Finally, check for a username and password

**Comment by bigdatasourav** [comment]
Hey @rinzool, Could you please address the above review comments?

File: databricks/service.go (hunk 3 above)
Comment (rinzool) [reply databricks/service.go:130]: IMHO we should check for ClientID/Secret before user and password yes.
I wonder if it should even be before token.
OAuth is more secure than workspace token which is more secure than user/password.  So should we check in that order?

Comment (bigdatasourav) [reply databricks/service.go:130]: Yes, then the order should be like the one below - 
// Account ID is required for a common column
// Check for profile and config file first
// If not using a profile and config file, check for oauth
// If not using oauth, check for token
// Finally, check for a username and password

**Comment by rinzool** [comment]
@bigdatasourav What do you think about https://github.com/turbot/steampipe-plugin-databricks/pull/6/commits/3e5e511fa8bef7f0cdec48fd381eefe408e30545?
I update the order and add checks before using token/user/password

File: databricks/service.go [hunk 4]
```diff
@@ -52,12 +52,19 @@ func getAccountClientUncached(ctx context.Context, d *plugin.QueryData, h *plugi
...
-		if databricksConfig.AccountToken != nil {
-			config.Token = *databricksConfig.AccountToken
+		if databricksConfig.ClientID != nil && databricksConfig.ClientSecret != nil {
+			config.ClientID = *databricksConfig.ClientID
+			config.ClientSecret = *databricksConfig.ClientSecret
+		}
+
+		empty_oauth_config := config.ClientID == "" && config.ClientSecret == "" && os.Getenv("DATABRICKS_CLIENT_ID") == "" && os.Getenv("DATABRICKS_CLIENT_SECRET") == ""
+
+		if empty_oauth_config && databricksConfig.WorkspaceToken != nil {
```
Comment (bigdatasourav) [inline databricks/service.go:62]:
```suggestion

                // If not using OAuth config, check for token
		if empty_oauth_config && databricksConfig.WorkspaceToken != nil {
```

Comment (bigdatasourav) [reply databricks/service.go:62]: Similarly, line number 47 also needs to be updated - `If not using a profile and config file, check for OAuth config`

Comment (rinzool) [reply databricks/service.go:63]: You're right. It's commited :+1:

Comment (bigdatasourav) [reply databricks/service.go:63]: Line number 47 is still not updated.

Comment (rinzool) [reply databricks/service.go:63]: Oh yes I forgot. It's done

File: README.md [hunk 5]
```diff
@@ -80,6 +88,8 @@ export DATABRICKS_HOST=https://accounts.cloud.databricks.com
 export DATABRICKS_ACCOUNT_ID=abcdd0f81-9be0-4425-9e29-3a7d96782373
 export DATABRICKS_USERNAME=user@turbot.com
 export DATABRICKS_PASSWORD=password
+export DATABRICKS_CLIENT_ID=client_id
```
Comment (bigdatasourav) [inline README.md:91]:
```suggestion
export DATABRICKS_CLIENT_ID=123-456-789
```

File: README.md [hunk 6]
```diff
@@ -80,6 +88,8 @@ export DATABRICKS_HOST=https://accounts.cloud.databricks.com
...
+export DATABRICKS_CLIENT_SECRET=client_secret
```
Comment (bigdatasourav) [inline README.md:92]:
```suggestion
export DATABRICKS_CLIENT_SECRET=dose1234567789abcde
```

File: databricks/service.go [hunk 7]
```diff
@@ -44,20 +44,28 @@ func getAccountClientUncached(ctx context.Context, d *plugin.QueryData, h *plugi
...
+		if databricksConfig.ClientID != nil && databricksConfig.ClientSecret != nil {
+			config.ClientID = *databricksConfig.ClientID
+			config.ClientSecret = *databricksConfig.ClientSecret
+		}
+
+		empty_oauth_config := config.ClientID == "" && config.ClientSecret == "" && os.Getenv("DATABRICKS_CLIENT_ID") == "" && os.Getenv("DATABRICKS_CLIENT_SECRET") == ""
+
+                // If not using OAuth config, check for token
+		if empty_oauth_config && databricksConfig.WorkspaceToken != nil {
+			config.Token = *databricksConfig.WorkspaceToken
```
Comment (bigdatasourav) [inline databricks/service.go:64]:
```suggestion
			config.Token = *databricksConfig.AccountToken
```

File: docs/index.md [hunk 8]
```diff
@@ -290,6 +298,23 @@ connection "databricks_user1-workspace" {
...
+
+  client_id = "123-456-789"
+  client_secret = "dose1234567789abcde"
+}
+```
+
 ### Credentials from Environment Variables
 
 Alternatively, you can also use the standard Databricks environment variables to obtain credentials **only if other argument (`profile`, `account_id`, `account_token`/`account_host`/`workspace_token`/`workspace_host`) is not specified** in the connection:
```
Comment (bigdatasourav) [inline docs/index.md:320]:
```suggestion
Alternatively, you can also use the standard Databricks environment variables to obtain credentials **only if other argument (`profile`, `account_id`, `client_id`/`client_secret`/`account_host`/`workspace_host`, `account_token`/`account_host`/`workspace_token`/`workspace_host`) is not specified** in the connection:
```

File: databricks/service.go [hunk 9]
```diff
@@ -44,20 +44,28 @@ func getAccountClientUncached(ctx context.Context, d *plugin.QueryData, h *plugi
...
-			config.Token = *databricksConfig.AccountToken
+		if databricksConfig.ClientID != nil && databricksConfig.ClientSecret != nil {
+			config.ClientID = *databricksConfig.ClientID
+			config.ClientSecret = *databricksConfig.ClientSecret
+		}
+
+		empty_oauth_config := config.ClientID == "" && config.ClientSecret == "" && os.Getenv("DATABRICKS_CLIENT_ID") == "" && os.Getenv("DATABRICKS_CLIENT_SECRET") == ""
+
+                // If not using OAuth config, check for token
+		if empty_oauth_config && databricksConfig.WorkspaceToken != nil {
```
[the remark this answers is not in the document]
Comment (rinzool) [reply databricks/service.go:63]: Thank you for spotting that error!

**Comment by rinzool** [comment]
@bigdatasourav Thanks for all your comments. I think it should be all good now

**Review by bigdatasourav — APPROVED** [review]

**Pull request merged**